src/stacked_borrows.rs

```diff
@@ -111,7 +118,12 @@ pub struct GlobalStateInner {
     tracked_call_ids: HashSet<CallId>,
     /// Whether to track raw pointers.
     tag_raw: bool,
+    /// Extra per-allocation information
+    extras: HashMap<AllocId, AllocHistory>,
```

Comment:
Why is this not in `AllocExtra` (aka `Stacks`)?

Reply:
Well now it is because we need the local crate list inside `stacked_borrows/diagnostics.rs`. I could add another parameter to the `log_*` functions so that we can pass the other piece of nonlocal state they need. But actually we only need those two parameters in combination. Do you think it would be reasonable to build a `Fn` (or some similar type) in the `Machine` and pass that instead of the `TagHistory` down to the diagnostics code?

Reply:
I am confused, what does this `extras` thing here have to do with the local crate list...?

I am also not sure what you mean by passing around a `Fn`.

Reply:
I think we should at least try to move `extras` into `Stacks`. It looks to me like the `log_*` functions could actually be implemented on `Stacks`.

Reply:
We need access to the list of local crates to decide what the "current span" is when we log an event. We only ever need the `ThreadManager` and the list of local crates at the same time, to call `current_span` where we walk the current call stack for a span that's in a local crate.

Reply:
(I'm moving the `AllocHistory` structure into `Stacks`)

Reply:
Ah so you want a `Fn` that, when invoked, returns the `Span`, and the `Fn` can capture internally the ThreadManager and the local crates list?

I guess that could work. So that'd be a `get_span: impl Fn() -> Span` everywhere?

Reply:
I think I'm fine with the latest refactor, we can fine tune in separate PRs

Reply:
Now we have an `Rc` in each `Stack`, isn't that bad for performance?
OTOH, that bit of refcount arithmetic is probably dwarfed by the memory allocator traffic for the `Vec`, so I guess we are good.